File: Year-2019/Mar/2019-03-22-215/lb/215.java
public class FindKthLargest {
    //堆排序
    public int findKthLargest(int[] nums, int k) {
        for (int i = nums.length/2-1;i>=0;i--)
            ajustHead(nums,i,nums.length);
        for (int i = nums.length-1;i>=nums.length-k;i--){
            swap(nums,0,i);
            ajustHead(nums,0,i);
        }
        return nums[nums.length-k];
    }

    public void ajustHead(int[] nums,int i,int length){
        for (int j = 2*i+1;j<length;j=2*j+1){
            if(j+1<length && nums[j+1]>nums[j])
                j++;
            if (nums[j]>nums[i]){
                swap(nums,i,j);
                i = j;
            }else{
                break;
            }
        }
    }

    public void swap(int[] nums,int i,int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: Year-2019/Mar/2019-03-22-215/watercow/215.c
// 利用快排partition函数做到O(n)复杂度
class Solution {
public:
	int findKthLargest(vector<int>& nums, int k) {
		int left = 0, right = nums.size() - 1;
		return __findKthLargest(nums, left, right, k);
	}

	int __findKthLargest(vector<int>& nums, int left, int right, int k) {
		int index = __partition(nums, left, right);
		if (index == (k - 1)) {
			return nums[index];
		}
		else if (index > k - 1) {
			return __findKthLargest(nums, left, index - 1, k);
		}
		else {
			return __findKthLargest(nums, index + 1, right, k);
		}
	}

	int __partition(vector<int>& nums, int left, int right) {
		// 返回当前访问元素排序后的index值
		int v = nums[left]; 

		// nums[l+1,...j] < v 且 nums[j+1, ...,r] > v
		int j = left;
		for (int i = j + 1; i <= right; i++) {
			if (nums[i] > v) {
				swap(nums[++j], nums[i]);
			}
		}

		swap(nums[left], nums[j]);
		return j;
	}
};